Sources/ArcGISToolkit/FloorFilter/FloorFilterViewModel.swift

```diff
@@ -0,0 +1,126 @@
+    }
+    
+    /// Facilities in the selected site
+    /// If no site is selected then the list is empty
+    /// If the sites data does not exist in the map, then use all the facilities in the map
+    public var facilities: [AGSFloorFacility] {
+        guard let floorManager = floorManager else { return [] }
+        return sites.isEmpty ? floorManager.facilities : floorManager.facilities.filter { $0.site == selectedSite }
+    }
+    
+    /// Levels that are visible in the expanded Floor Filter levels table view
+    /// Reverse the order of the levels to make it in ascending order
+    public var visibleLevelsInExpandedList: [AGSFloorLevel] {
+        guard let floorManager = floorManager else { return [] }
+        return facilities.isEmpty ? floorManager.levels : floorManager.levels.filter { $0.facility == selectedFacility }.reversed()
+    }
+    
+    /// All the levels in the map
+    private var allLevels: [AGSFloorLevel] {
+        return floorManager?.levels ?? []
+    }
+    
+    /// The site, facility, and level that are selected by the user
+    public var selectedSite: AGSFloorSite?
+    public var selectedFacility: AGSFloorFacility?
+    public var selectedLevel: AGSFloorLevel?
+    
+    /// The default vertical order is 0 according to Runtime 100.12 update for AGSFloorManager
+    static private let defaultVerticalOrder = 0
+
+    /// Reset FloorFilter saved data
+    public func reset() {
+        floorManager = nil
+        selectedSite = nil
+        selectedFacility = nil
+        selectedLevel = nil
+    }
+    
+    /// Gets the default level for a facility
+    /// Uses level with vertical order 0 otherwise gets the lowest level
+    public func getDefaultLevelForFacility(facility: AGSFloorFacility?) -> AGSFloorLevel? {
+        let candidateLevels = allLevels.filter { $0.facility == facility }
+        return candidateLevels.first { $0.verticalOrder == 0 } ?? getLowestLevel(levels: candidateLevels)
+    }
+    
+    /// Returns the AGSFloorLevel with the lowest verticalOrder.
+    private func getLowestLevel(levels: [AGSFloorLevel]) -> AGSFloorLevel? {
+        let sortedLevels = allLevels.sorted {
+            $0.verticalOrder < $1.verticalOrder
+        }
+        return sortedLevels.first {
+            $0.verticalOrder != Int.min && $0.verticalOrder != Int.max
+        }
+    }
+    
+    /// Sets the visibility of all the levels on the map based on the vertical order of the current selected level
+    public func filterMapToSelectedLevel() {
+        guard let selectedLevel = selectedLevel else { return }
+        allLevels.forEach {
+            $0.isVisible = $0.verticalOrder == selectedLevel.verticalOrder
+        }
+    }
+    
+    /// Zooms to the facility if there is a selected facility, otherwise zooms to the site.
+    public func zoomToSelection() {
+        if selectedFacility != nil {
+            zoomToFacility()
+        } else if selectedSite != nil {
+            zoomToSite()
+        }
```

Comment:
Both `selectedFacility` and `selectedSite` are checked to be non-nil, but then they are checked again in the corresponding method using optional chaining. Instead, it would be better to unwrap them here using `if let` and then pass the unwrapped facility or site to the method:

```swift
public func zoomToSelection() {
    if let selectedFacility = selectedFacility {
        zoom(to: selectedFacility)
    } else if let selectedSite = selectedSite {
        zoom(to: selectedSite)
    }
}

private func zoom(to floorSite: AGSFloorSite) {
    zoomToExtent(mapView: mapView, envelope: floorSite.geometry?.extent)
}

private func zoom(to floorFacility: AGSFloorFacility) {
    zoomToExtent(mapView: mapView, envelope: floorFacility.geometry?.extent)
}
```